src/Persistent_cohomology/include/gudhi/Persistent_cohomology/Field_Zp.h

```diff
@@ -34,15 +35,34 @@ class Field_Zp {
 
   void init(int charac) {
     assert(charac > 0);  // division by zero + non negative values
+
     Prime = charac;
+
+    // Check that the provided prime is less than the maximum allowed as int and calculation below : 46337 ; i.e (max_prime-1)**2 <= INT_MAX
+    if(Prime > 46337)
+        throw std::invalid_argument("Maximum homology_coeff_field allowed value is 46337");
+
+    // Check for primality
+    if ((Prime == 0) || (Prime == 1) || ((Prime > 3) && ((Prime % 2 == 0) || (Prime % 3 == 0))))
+        throw std::invalid_argument("homology_coeff_field must be a prime number");
+
     inverse_.clear();
     inverse_.reserve(charac);
     inverse_.push_back(0);
     for (int i = 1; i < Prime; ++i) {
       int inv = 1;
-      while (((inv * i) % Prime) != 1)
+      int mult = inv * i;
+      while ( (mult % Prime) != 1) {
         ++inv;
+        if(mult == Prime)
+            throw std::invalid_argument("homology_coeff_field must be a prime number");
+        mult = inv * i;
+      }
       inverse_.push_back(inv);
+      if ( (i*i <= Prime) && (((i-5)%6) == 0) ) {
+        if ((Prime % i == 0) || (Prime % (i + 2) == 0))
+            throw std::invalid_argument("homology_coeff_field must be a prime number");
```

Comment:
I don't think this extra test is necessary, the test `mult==Prime` above seems sufficient. If Prime is not a prime, we will first have some `i` that are coprime with it so we can find an inverse, and then `i` will be the smallest prime factor and we will reach `inv=Prime/i` and detect the issue.